[feat] 상품 삭제 기능 구현

pie2457 opened the pull request:
## Issues
- #24 

## What is this PR? 👓
- 상품 삭제 기능에 대한 PR입니다.

## Key changes 🔑 
- 상품 삭제 기능
- 상품 삭제 테스트 작성

##  📖 공부하면서 배웠던 것
### `@Modifying` 이란 ?
- `@Query` 어노테이션(JPQL Query, Native Query)을 통해 작성된 INSERT, UPDATE, DELETE(SELECT 제외) 쿼리에서 사용되는 어노테이션입니다.
  - `@Query`를 사용시 `@Modifying` 애노테이션을 붙여주지 않으면 `QueryExecutionRequestException`이 발생한다.
- __기본적으로 JpaRepository에서 제공하는 메서드 혹은 메서드 네이밍으로 만들어진 쿼리에는 적용되지 않습니다__.
- clearAutomatically, flushAutomatically 속성을 변경 할 수 있으며 **주로 벌크 연산과 같이 이용**됩니다.
  - clearAutomatically ( default: false )
  - flushAutomatically ( default: false )
### 📌 clearAutomatically ?
JPA에서 1차 캐시는 DB의 접근 횟수를 줄여주고 다양한 성능 개선의 효과를 가져다주지만 `@Modifying`과 `@Query`를 이용한 벌크 연산에서는 이 기능때문에 예측 불가능한 결과가 나올 수 있다.
JPA에서 조회를 실행할 때 1차 캐시를 확인해서 해당 엔티티가 1차 캐시에 존재한다면 DB에 접근하지 않고, 1차캐시의 엔티티를 반환한다. 하지만 **벌크연산은 1차 캐시를 포함한 영속성 컨텍스트를 무시하고 바로 QUERY를 실행하기 때문에 영속성 컨텍스트는 데이터 변경을 알 수 없다. 즉, 벌크연산을 실행할 때 싱크가 맞지 않게된다**.
이 문제를 해결하기위해 `@Modifying`어노테이션은 **ClearAutomatically**라는 속성이 존재하는 것이다. defalt가 False인 이 속성을 true로 변경해준다면 **벌크연산 직후 자동으로 영속성 컨텍스트를 clear해준다**. 영속성 컨텍스트를 clear해준다면 조회를 실행할 때 이미 update결과가 진행되어진 DB를 조회하기 때문에 데이터의 동기화문제를 해결할 수 있다.
> 한줄 요약
`@Query` 쿼리 실행 후에 영속성 컨텍스트를 비워서 DB에서 수정된 값을 정상적으로 조회할 수 있게 하기 위해 사용한다.
### 📌 flushAutomatically ?
`clearAutomatically`는 JPQL을 실행한 후 동작하지만 `flushAutomatically` JPQL을 DB에 날리기 전 영속성 컨텍스트를 flush해준다.
여기서 한가지 의문이 생겼다. 트랜잭션 내에서 JPQL를 실행하면, 영속성 컨텍스트가 flush 된다고 알고 있는데, `flushAutomatically` 옵션은 왜 생긴것인지 궁금해 찾아보았다.

> AUTO flush
By default, Hibernate uses the AUTO flush mode which triggers a flush in the following circumstances:
prior to committing a Transaction
prior to executing a JPQL/HQL query that overlaps with the queued entity actions
before executing any native SQL query that has no registered synchronization

내가 이해하기로는 실행될 JPQL에 의해서 영향을 받는 Entity들에 대해서만 flush가 된다는 것이었다.
(물론 이건 FlushMode 옵션이 기본값인 FlushModeType.AUTO 일 때)

간단한 로직을 통해 더 쉽게 이해해보자.
~~~java
@Transactional
	public void update(Long productId) {
		Product product = productRepository.findProductById(productId);
		
		product.changePrice("10000");  // 1번
		bookmarkRepository.deleteByProductId(product.getId());  // 2번
                // deleteByProductId()는 clearAutomatically=true 옵션을 설정해놨다.
	}
~~~~
1. 상품의 가격을 변경 시킨다.
2. 북마크로 설정해놓은 아이템을 해지 시킨다.

위의 경우 product 가격을 10000원으로 변경하는 쿼리는 메서드가 끝날 때, Dirty Checking(변경 감지)에 의해서 Transaction이 커밋되면서 flush 돼야한다. 하지만 해당 쿼리는 어디서도 발생하지 않는다. 왜냐하면  Hibernate가 deleteByProductId()의 쿼리가 product에는 영향을 미치지 않는다고 판단하여 product.changePrice() 쿼리를 flush를 시키지 않고, deleteByProductId() 호출하며`clearAutomatically=true` 옵션에 의해서 영속성 컨텍스트가 싹 비워지기 때문이다.
즉, 쓰기 지연 저장소에 저장되어 있던 product.changePrice() 쿼리도 삭제되어 버렸다. 이러한 문제점들로 인해 `flushAutomatically` 옵션을 사용한다.

> 한줄 요약
@Query 쿼리 실행 전에 쓰기 지연 저장소에 남아있는 쿼리를 미리 flush 하고자 할 때 사용한다

[reference] 
- https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#flushing-auto
- https://velog.io/@gruzzimo/JPA-Modifying%EC%9D%98-flushAutomatically-%EC%98%B5%EC%85%98%EC%9D%80-%EC%96%B8%EC%A0%9C-%EC%93%B0%EC%A7%80

pie2457 committed:
feat #24 : 상품 삭제 기능

pie2457 committed:
test #24 : 상품 삭제 기능 테스트

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/item/ItemService.java:
```diff
@@ -139,6 +143,26 @@ private Category findCategory(ItemModifyRequest modifyRequest) {
+
+		imageService.deleteImageFromS3(findItem.getThumbnail());
+		deleteAllRelatedItem(itemId, principal.getMemberId());
+
```
빈 라인이 하나 더 들어갔네요~

pie2457 replied:
앗 네 수정하겠습니다 ~!

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/item/ItemService.java:
```diff
@@ -139,6 +143,26 @@ private Category findCategory(ItemModifyRequest modifyRequest) {
+	@Transactional
+	public void deleteItem(Long itemId, Principal principal) {
+		verifyExistsMember(principal.getMemberId());
+		verifyExistsItem(itemId);
+
+		Item findItem = itemRepository.findItemByIdAndMemberId(itemId, principal.getMemberId())
+			.orElseThrow(() -> new ForbiddenException(ErrorCode.ITEM_FORBIDDEN));
```
밑에서 없으면 orElseThrow 하는데 굳이 `verifyExistsItem`으로 있는지 검증할 필요가 있을까용?

pie2457 replied:
`verifyExistsItem`은 itemId에 해당하는 Item이 있는지 검증하는 로직이고 , 말씀하신 부분은 item은 있는데 itemId와 memberId가 일치하는 item이 없다면 접근 권한이 없는거라 생각해서 exception을 던지는 로직인데,
`findItemByIdAndMemberId`만으로는 item이 없는지 member가 없는지 itemId와 memberId에 해당하는 item이 없는지 어떤 에러를 던져줘야하는지 분간이 안갈거같아서요..!

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/persistence/ChatroomRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface ChatroomRepository extends JpaRepository<Chatroom, Long> {
+
+	void deleteByItemId(Long itemId);
```
하나의 상품에 연관된 채팅방이 여러 개가 되면 쿼리가 어떻게 나가게 될까요?

pie2457 replied:
여러번 나가게 되겠죠 ㅠㅠ? bulk delete도 가능한지 찾아봐야겠네욥.. 감사합니당 ~!

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/persistence/image/ImageRepository.java:
```diff
@@ -20,4 +20,6 @@ void deleteImageByItemIdAndImageUrls(@Param("itemId") Long itemId,
 	void deleteImageByItemIdAndImageUrl(@Param("itemId") Long itemId,
 		@Param("imageUrl") String imageUrl);
+
+	void deleteByItemId(Long itemId);
```
그러고보니 여기는 `@Modifying`을 붙이지 않았네요~? 이유가 있을까요?

pie2457 replied:
조건이 있을때만 쓰는건 줄 알았어요.. ! `@Modifying` 애노테이션에 대해 더 잘 알아보겠습니당 ㅠ

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/persistence/item/ItemRepository.java:
```diff
@@ -16,6 +16,8 @@ public interface ItemRepository extends JpaRepository<Item, Long> {
 
 	boolean existsItemById(Long itemId);
 
+	void deleteItemByIdAndMemberId(Long itemId, Long memberId);
```
`ItemRepository`에서 상품(Item)을 삭제하는데 `itemId` 면 충분하지 않나요? `memberId`를 추가로 넣으신 이유가 있을까용?

pie2457 replied:
memberId가 있으면 삭제하는 상품에 대해 더 명확할꺼같았는데, 생각해보니 ItemId 자체가 unique해서 itemId만 있어도 될것같네용 ㅎㅎ..

23Yong commented on src/test/java/pie/tomato/tomatomarket/application/ItemServiceTest.java:
```diff
@@ -247,6 +247,32 @@ void modifyItemVer_2() {
+		itemService.deleteItem(item.getId(), principal);
+
+		// then
+		assertAll(
+			() -> assertThat(supportRepository.findById(item.getId(), Item.class)).isNull(),
+			() -> assertThat(imageRepository.findById(item.getId())).isEmpty()
+		);
+
```
`.isNull()`과 `.isEmpty()`의 차이는 뭐에용?
273라인 빈 줄이 하나 더 들어갔어용~

pie2457 replied:
`isEmpty()`는 Optional 처리에 대한 반환값입니다 ! !
넵 수정 하겠습니다 ~!

23Yong reviewed (approved):
수고하셨어용~

pie2457 committed:
refactor #24 : 리뷰사항 반영

- ItemService : 불필요한 공백 제거 및 메서드명 변경
- ItemServiceTest : 불필요한 공백 제거
- ChatroomRepository : @Query를 사용해 delete 쿼리가 여러번 나가지 않도록 수정, @Modifying 애노테이션에 clearAutomatically, flushAutomatically 옵션 설정
- ImageRepository : @Modifying 애노테이션에 clearAutomatically, flushAutomatically 옵션 설정
- ItemRepository : 네이밍 메서드가 아닌 기본 CrudRepository를 사용하도록 변경

pie2457 committed:
chore #24 : 테스트 시 쿼리를 보기 위해 logging.level 설정
